File: pippip-native-ios/More/LocalPasswordCell.m
//
//  LocalPasswordCell.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 3/20/18.
//  Copyright © 2018 seComm. All rights reserved.
//

#import "LocalPasswordCell.h"
#import "pippip_native_ios-Swift.h"
#import "ApplicationSingleton.h"
#import "UserVault.h"

@interface LocalPasswordCell ()

@property (weak, nonatomic) IBOutlet UIButton *changePasswordButton;
@property (weak, nonatomic) IBOutlet UITextField *passphraseTextField;

@end

@implementation LocalPasswordCell

+ (MoreCellItem*)cellItem {

    MoreCellItem *item = [[MoreCellItem alloc] init];
    item.cellReuseId = @"LocalPasswordCell";
    item.cellHeight = 65.0;
    return item;

}

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    
    _changePasswordButton.alpha = 0.0;
    [_changePasswordButton setEnabled:NO];

}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}
/*
- (void)setViewController:(UIViewController *)view {
    _moreView = view;
}
*/
- (void)checkPassphrase:(NSString*)passphrase {

    if (_passphraseTextField.text.length > 0) {
        [self doChangePassphrase:passphrase];
    }
    else {

        NSString *message = @"Empty passphrases are not recommended.\nTap 'Ok' to continue, 'Cancel' to start over";
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Empty Passphrase"
                                                                       message:message
                                                                preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"Ok"
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction *action){
                                                             [self doChangePassphrase:passphrase];
                                                         }];
        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                               style:UIAlertActionStyleCancel
                                                             handler:^(UIAlertAction *action) {
                                                                 _passphraseTextField.text = @"***********";
                                                                 _changePasswordButton.alpha = 0.0;
                                                                 [_changePasswordButton setEnabled:NO];
                                                             }];
        [alert addAction:okAction];
        [alert addAction:cancelAction];
        NSMutableDictionary *info = [NSMutableDictionary dictionary];
        info[@"alert"] = alert;
        [[NSNotificationCenter defaultCenter] postNotificationName:@"PresentAlert" object:nil userInfo:info];

    }

}

- (void)doChangePassphrase:(NSString*)passphrase {

    SessionState *sessionState = [ApplicationSingleton instance].accountSession.sessionState;
    NSString *accountName = sessionState.currentAccount;
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSString *vaultPath = [vaultsPath stringByAppendingPathComponent:accountName];
    NSData *vaultData = [NSData dataWithContentsOfFile:vaultPath];

    UserVault *vault = [[UserVault alloc] initWithState:sessionState];
    NSError *error = nil;
    [vault decode:vaultData withPassword:passphrase withError:&error];
    if (error == nil) {
        [self storeVault:sessionState withPassphrase:_passphraseTextField.text];
    }
    else {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Invalid Passphrase"
                                                                       message:@""
                                                                preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                           style:UIAlertActionStyleDefault
                                                         handler:nil];
        [alert addAction:okAction];
        NSMutableDictionary *info = [NSMutableDictionary dictionary];
        info[@"alert"] = alert;
        [[NSNotificationCenter defaultCenter] postNotificationName:@"PresentAlert" object:nil userInfo:info];
        _passphraseTextField.text = @"***********";
        _changePasswordButton.alpha = 0.0;
        [_changePasswordButton setEnabled:NO];
    }

}

- (void)storeVault:(SessionState*)sessionState withPassphrase:(NSString*)passphrase {
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSString *vaultPath = [vaultsPath stringByAppendingPathComponent:sessionState.currentAccount];
    
    UserVault *vault = [[UserVault alloc] initWithState:sessionState];
    NSData *vaultData = [vault encode:passphrase];
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    [fileManager createDirectoryAtPath:vaultsPath
           withIntermediateDirectories:NO
                            attributes:nil
                                 error:nil];
    [vaultData writeToFile:vaultPath atomically:YES];
    _passphraseTextField.text = @"***********";
    _changePasswordButton.alpha = 0.0;
    [_changePasswordButton setEnabled:NO];

}

- (IBAction)passwordChanged:(UITextField *)sender {

    _changePasswordButton.alpha = 1.0;
    [_changePasswordButton setEnabled:YES];
    
}

- (IBAction)changePassword:(id)sender {

    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Change Local Passphrase"
                                                                   message:@"Enter your current passphrase" preferredStyle:UIAlertControllerStyleAlert];

    [alert addTextFieldWithConfigurationHandler:^(UITextField* textField) {
        textField.placeholder = @"Passphrase";
        textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    }];
    
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                       style:UIAlertActionStyleDefault
                                                     handler:^(UIAlertAction *action){
                                                         NSString *passphrase = alert.textFields[0].text;
                                                         [self checkPassphrase:passphrase];
                                                     }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:^(UIAlertAction *action){
                                                             _passphraseTextField.text = @"***********";
                                                             _changePasswordButton.alpha = 0.0;
                                                             [_changePasswordButton setEnabled:NO];
                                                         }];
    [alert addAction:okAction];
    [alert addAction:cancelAction];
    NSMutableDictionary *info = [NSMutableDictionary dictionary];
    info[@"alert"] = alert;
    [[NSNotificationCenter defaultCenter] postNotificationName:@"PresentAlert" object:nil userInfo:info];

}

@end

File: pippip-native-ios/More/LocalPasswordCell.swift
//
//  LocalPasswordCell.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 4/9/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import PMAlertController
import RKDropdownAlert
import ChameleonFramework

class LocalPasswordCell: TableViewCellWithController {

    @IBOutlet weak var passphraseText: UITextField!
    @IBOutlet weak var changePassphraseButton: UIButton!

    var sessionState = SessionState()

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code

        changePassphraseButton.isEnabled = false
        changePassphraseButton.alpha = 0.0
        
    }
    @objc class func cellItem() -> MoreCellItem {
        
        let item = MoreCellItem()
        item.cellHeight = 65.0
        item.cellReuseId = "LocalPasswordCell"
        return item
        
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    func doChangePassphrase(oldPassphrase: String, newPassphrase: String) {

        let vault = UserVault()
        do {
            try vault.changePassphrase(oldPassphrase: oldPassphrase, newPassphrase: newPassphrase)
            let alertColor = UIColor.flatLime
            RKDropdownAlert.title("Passphrase Changed", message: "Your local passphrase has been changed",
                                  backgroundColor: alertColor,
                                  textColor: ContrastColorOf(alertColor, returnFlat: true),
                                  time: 2, delegate: nil)
            resetCell()
        }
        catch {
            var info = [AnyHashable: Any]()
            info["title"] = "Change Passphrase Exception"
            info["message"] = "An error has occurred, please try again"
            NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
        }
    }

    func emptyPassphrase (oldPassphrase: String, newPassphrase: String) {

        DispatchQueue.main.async {
            let alert = PMAlertController(title: "Change Passphrase",
                                          description: "Empty passphrases are not recommended\nProceed?",
                                          image: nil,
                                          style: .alert)
            alert.addAction(PMAlertAction(title: "Yes", style: .default, action: { () in
                self.doChangePassphrase(oldPassphrase: oldPassphrase, newPassphrase: newPassphrase)
            }))
            alert.addAction(PMAlertAction(title: "No", style: .cancel, action: { () in
                self.resetCell()
            }))
            self.viewController?.present(alert, animated: true, completion: nil)
        }

    }

    func newPassphrase(_ oldPassphrase: String) {

        let passphrase = passphraseText.text!
        if passphrase.utf8.count == 0 {
            self.emptyPassphrase(oldPassphrase: oldPassphrase, newPassphrase: passphrase)
        }
        else {
            self.doChangePassphrase(oldPassphrase: oldPassphrase, newPassphrase: passphrase)
        }
    
    }

    func resetCell() {

        passphraseText.text = "***********"
        changePassphraseButton.alpha = 0.0
        changePassphraseButton.isEnabled = false

    }
    
    @IBAction func changePassphrase(_ sender: Any) {

        let alert = PMAlertController(title: "Change Passphrase",
                                      description: "Enter your current passphrase",
                                      image: nil,
                                      style: .alert)
        alert.addTextField({ (textField) in
            textField?.placeholder = "Passphrase"
            textField?.autocorrectionType = .no
            textField?.spellCheckingType = .no
            textField?.autocapitalizationType = .none
            textField?.becomeFirstResponder()
        })
        alert.addAction(PMAlertAction(title: "Verify Passphrase", style: .default, action: { () in
            let passphrase = alert.textFields[0].text ?? ""
            do {
                if try UserVault.validatePassphrase(passphrase) {
                    self.newPassphrase(passphrase)
                }
                else {
                    var info = [AnyHashable: Any]()
                    info["title"] = "Verify Passphrase"
                    info["message"] = "Invalid passphrase"
                    NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
                    self.resetCell()
                }
            }
            catch {
                var info = [AnyHashable: Any]()
                info["title"] = "Verify Passphrase Exception"
                info["message"] = "An error has occurred, please try again"
                NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
            }
        }))
        alert.addAction(PMAlertAction(title: "Cancel", style: .cancel, action: { () in
            self.resetCell()
        }))
        viewController?.present(alert, animated: true, completion: nil)

    }

    @IBAction func passphraseChanged(_ sender: Any) {

        changePassphraseButton.alpha = 1.0
        changePassphraseButton.isEnabled = true

    }

}
